src/Parsing/Impl/Ast/ScopeStatement.cs

```diff
@@ -15,274 +15,43 @@
-                        if (parent.TryBindOuter(this, variableName.Name, false, out var variable)) {
-                            bound = !variable.IsGlobal;
-                            break;
-                        }
-                    }
-
-                    if (!bound) {
-                        binder.ReportSyntaxError("no binding for nonlocal '{0}' found".FormatUI(variableName.Name), variableName);
-                    }
-                }
-            }
-
-            if (FreeVariables != null && FreeVariables.Count > 0) {
-                closureVariables = new List<ClosureInfo>();
-
-                foreach (var variable in FreeVariables) {
-                    closureVariables.Add(new ClosureInfo(variable, !(this is ClassDefinition)));
-                }
-            }
-
-            if (Variables != null && Variables.Count > 0) {
-                if (closureVariables == null) {
-                    closureVariables = new List<ClosureInfo>();
-                }
-
-                foreach (var variable in Variables.Values) {
-                    if (!HasClosureVariable(closureVariables, variable) &&
-                        !variable.IsGlobal && (variable.AccessedInNestedScope || ExposesLocalVariable(variable))) {
-                        closureVariables.Add(new ClosureInfo(variable, true));
-                    }
+        public ScopeStatement Parent => this.FindClosestScopeStatement();
 
-                    if (variable.Kind == VariableKind.Local) {
-                        Debug.Assert(variable.Scope == this);
-                    }
-                }
-            }
+        public IScopeNode ParentNode { get; set; }
 
-            // no longer needed
-            _references = null;
-        }
+        public virtual Statement Body { get; }
 
-        private static bool HasClosureVariable(List<ClosureInfo> closureVariables, PythonVariable variable) {
-            if (closureVariables == null) {
-                return false;
-            }
+        public bool IsClosure => ScopeInfo.IsClosure;
 
-            for (var i = 0; i < closureVariables.Count; i++) {
-                if (closureVariables[i].Variable == variable) {
-                    return true;
-                }
-            }
+        public bool ContainsNestedFreeVariables => ScopeInfo.ContainsNestedFreeVariables;
 
-            return false;
-        }
+        public bool NeedsLocalsDictionary => ScopeInfo.NeedsLocalsDictionary;
 
-        private void EnsureVariables() {
-            if (Variables == null) {
-                Variables = new Dictionary<string, PythonVariable>(StringComparer.Ordinal);
-            }
-        }
+        public virtual string Name => "<unknown>";
 
-        internal void AddVariable(PythonVariable variable) {
-            EnsureVariables();
-            Variables[variable.Name] = variable;
-        }
+        public ICollection<PythonVariable> ScopeVariables => ScopeInfo.ScopeVariables;
```

Comment:
Should be readonly collection, such as `ImmutableArray` or at least `IReadOnlyCollection`, we don't want to expose `Remove(T)`

Reply:
Yes!